### ISSUE TITLE
error unregistering CLI instance with Hammerspoon: error during transport

### ISSUE THREAD

**Issue by bradyt** [Hammerspoon/hammerspoon#2038]
If my `init.lua` only contains `hs.ipc.cliInstall()`, and I run `hs -c 'hs.reload()'` in terminal, I get the following message in terminal.

```
error unregistering CLI instance with Hammerspoon: error during transport
```

This behavior has been confirmed by another user on IRC.

The error does not occur with less trivial `init.lua`.

**Comment by Habbie** [comment]
> This behavior has been confirmed by another user on IRC.

That was me.

**Comment by asmagill** [comment]
Apparently there is a dependency on another module that isn't being loaded somewhere... changing `init.lua` from:

~~~lua
hs.ipc.cliInstall()
~~~

to

~~~lua
require("hs.window")
hs.ipc.cliInstall()
~~~

doesn't cause the error to occur.

In either case, even when the error message is returned, Hamemrspoon *is* reloading.

Since I contributed most of the re-write to the `hs.ipc` module, I'll see if I can find some time this weekend to track it down further... I've been away from contributing to the development of Hamemrspoon for awhile now and I'd like to try and get back into it... this may give me the push I need 😛

**Comment by latenitefilms** [comment]
Maybe `hs.doc.hsdocs`?

**Comment by asmagill** [comment]
Ok, to the best of my ability to determine, this error message occurs because Hammerspoon is restarting too quickly and the message port is torn down in the Hammerspoon thread before the `sendToRemote:` method in the `hs` thread returns... having something else that has a lot of background stuff like `hs.window` loaded slows the Hammerspoon thread down enough that `hs` ends before the message port is actually terminated, but even this isn't 100% (I did get this message once even with `hs.window` loaded).

This happens because the function call `hs.reload` doesn't actually return, so `sendToRemote:` has to time out (even when we tell it not to expect a response, apparently it can sometimes take longer to return than something that acknowledges immediately)... the Lua state is forced into final garbage collection and a brand new one is created... so adding a timer to delay the teardown in the `init.lua` file doesn't help (it's never executed) and adding it to the garbage collection in `internal.m` causes the message port to still be in existence when Hammerspoon restarts so loading `hs.ipc` fails causing `init.lua` not to complete, usually followed by a crash (not entirely sure why; it's in `objc_retain` but I'm pretty sure I balanced that correctly in my test code... however since it prevents `hs.ipc` from reloading on restart in any case, I'm not digging into that for now).

So, unless someone perusing this has experience with `CFMessagePort` and wants to chime in with something I'm missing, the simplest solution is to ignore/suppress the message, e.g. something like `hs -c "hs.reload()" 2> /dev/null`. I could remove the message entirely from the `hs` code, but since it *can* be suppressed at the shell level, I'd rather not unless there is strong consensus otherwise... I'm not a fan of commands that don't give error messages as it makes troubleshooting things more difficult and it's usually not too hard to hide them if it can be determined that the message isn't actually important (as in this case).
